### PR TITLE
Add testutil package

### PR DESCRIPTION
**Pull request by beorn7**
This is derived from #423. I recreated it so that it is pulling from a branch I can push to...

Changes to #423 are based on the review that has already happened. I packaged it nicely in commits with messages, so you can follow along.

I will add more documentation and the single-value retrieval helpers (cf. #230) in a separate PR. Same with the creation of an internal package for the duplicated code. Don't want to put too many things into this one PR.

@grobie @brancz

### CODE REVIEW

**Comment by beorn7** [comment]
@brancz @grobie could you have a look?

File: prometheus/testutil/testutil.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+	var filtered []*dto.MetricFamily
+	for _, m := range metrics {
+		drop := true
+		for _, name := range names {
+			if m.GetName() == name {
+				drop = false
+				break
+			}
+		}
+		if !drop {
```
Comment (cbandy) [inline prometheus/testutil/testutil.go]: These loops make more sense to me without any boolean var:

```go
if m.GetName() == name {
  filtered = append(filtered, m)
  break
}
```

File: prometheus/testutil/testutil.go [hunk 2]
```diff
@@ -26,29 +27,37 @@ import (
...
 )
 
-// GatherAndCompare retrieves all metrics exposed by a collector and compares it
-// to an expected output in the Prometheus text exposition format.
-// metricNames allows only comparing the given metrics. All are compared if it's nil.
-func GatherAndCompare(c prometheus.Collector, expected string, metricNames ...string) error {
+// CollectAndCompare registers the provided Collector with a newly created
+// pedantic Registry. It then does the same as GatherAndCompare, gathering the
+// metrics from the pedantic Registry.
+func CollectAndCompare(c prometheus.Collector, expected io.Reader, metricNames ...string) error {
```
[the remark this answers is not in the document]
Comment (beorn7) [reply prometheus/testutil/testutil.go]: Exactly. I will fix the commit description.

File: prometheus/testutil/testutil.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+func GatherAndCompare(g prometheus.Gatherer, expected io.Reader, metricNames ...string) error {
+	metrics, err := g.Gather()
+	if err != nil {
+		return fmt.Errorf("gathering metrics failed: %s", err)
+	}
+	if metricNames != nil {
+		metrics = filterMetrics(metrics, metricNames)
+	}
+	var tp expfmt.TextParser
+	expectedMetrics, err := tp.TextToMetricFamilies(expected)
```
[the remark this answers is not in the document]
Comment (beorn7) [reply prometheus/testutil/testutil.go]: It's legal in the text format to liberally use whitespace. Knowing exactly the formatting of the encoder would be a burden for the programmer writing those tests, and one for which I cannot really see any gain.

Personally, I wouldn't even like if a "normalized" formatting is advertised (but that's a discussion that is happening elsewhere and definitely out of scope of this PR). In any case, there is no goal right now to provide a tool for "normalizing".

File: prometheus/testutil/testutil.go (hunk 1 above)
Comment (beorn7) [reply prometheus/testutil/testutil.go]: I think you are right.

NB: This is not my code. So I don't know if there is some non-obvious reason why it was done the way it is.

I'll change it according to your suggestion, and @brancz can chime in if we blundered here.

**Comment by beorn7** [comment]
@cbandy Thanks for your review. I incorporated the suggested changes (except the one requiring the test writer to follow a certain formatting, see my comment above).

Still, I would feel better to get a look from @grobie or @brancz before merging.

**Comment by beorn7** [comment]
That's all mapped out in the initial description of this PR. :o)

**Comment by grobie** [comment]
Looks good to me!

**Pull request merged**